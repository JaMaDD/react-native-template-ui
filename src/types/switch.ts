import type { ViewStyle } from 'react-native';
import type { SwitchSize } from '../utils/theme/const';
import type { AnimatedThemedPressableProps, OnPressDelayConfig } from './btn';
import type { CustomAnimation } from './reanimated';
import type { StyleOrStyleProp } from './style';
import type { ThemeColors } from './theme';
import type { AnimatedThemedViewProps } from './view';

export type CustomSwitchColors = {
  border?: ThemeColors;
  borderEnabled?: ThemeColors;
  background?: ThemeColors;
  backgroundEnabled?: ThemeColors;
  thumb?: ThemeColors;
  thumbEnabled?: ThemeColors;
};

export type ThemedSwitchProps = Omit<
  AnimatedThemedPressableProps,
  'onPress' | 'children'
> & {
  onPress: (enabled: boolean) => void;
  enabled?: boolean;
  size?: SwitchSize | number;
  onPressDelayConfig?: OnPressDelayConfig;
  customEnableAnimation?: CustomAnimation<1>;
  customDisableAnimation?: CustomAnimation<0>;
  customColors?: CustomSwitchColors;
  thumbProps?: AnimatedThemedViewProps;
  style?: StyleOrStyleProp<ViewStyle>;
};
