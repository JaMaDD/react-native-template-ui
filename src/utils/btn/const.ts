export enum BtnScaleRatio {
  Square = 0.9,
  Rectangle = 0.98,
}

export enum OnPressDelayType {
  Throttle,
  Debounce,
}
