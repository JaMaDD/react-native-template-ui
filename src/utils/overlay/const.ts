import { IconSize } from '@jamadd/react-native-template-icons';
import { createContext } from 'react';
import type {
  ActionSheetContextVal,
  AlertContextVal,
} from '../../types/overlay';
import type { ThemeSpacing, ThemeTextVariants } from '../../types/theme';

/** @internal */
export const overlayMaxWidthPercent = '80%';

export enum OverlayDismissResultType {
  Toast = 'toast',
  Alert = 'alert',
  ActionSheet = 'actionSheet',
}

export const overlayDismissResultDefaultText = 'cancelByUser';

export enum ToastType {
  Top = 'top',
  TopLeft = 'topLeft',
  TopRight = 'topRight',
  Bottom = 'bottom',
  BottomLeft = 'bottomLeft',
  BottomRight = 'bottomRight',
}

export enum ToastDuration {
  Short = 3000,
  Long = 5000,
}

/** @internal */
export const AlertContext = createContext<AlertContextVal>({});

export enum AlertBtnType {
  Primary = 'primary',
  Secondary = 'secondary',
  Destructive = 'destructive',
}

/** @internal */
export const ActionSheetContext = createContext<Partial<ActionSheetContextVal>>(
  {}
);

/** @internal */
export const actionSheetOpenDuration = 300;

/** @internal */
export const actionSheetSnapDuration = 300;

/** @internal */
export const actionSheetDismissDuration = 150;

/** @internal */
export const actionSheetMaxHeightRatio = 0.6;

/** @internal */
export enum ActionSheetExpandableHeightRatio {
  Top = 0.9,
  Middle = 0.6,
  Bottom = 0.3,
}

/** @internal */
export const actionSheetHeaderIconSize: IconSize = IconSize.M;

/** @internal */
export const actionSheetHeaderPadding: ThemeSpacing = 'm';

/** @internal */
export const actionSheetOptListItemPadding: ThemeSpacing = 'm';

/** @internal */
export const actionSheetOptListItemTextVariant: ThemeTextVariants = 'textM';
