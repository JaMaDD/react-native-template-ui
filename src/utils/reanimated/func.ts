import type {
  SharedValue,
  WithSpringConfig,
  WithTimingConfig,
} from 'react-native-reanimated';
import { withSpring, withTiming } from 'react-native-reanimated';
import { scheduleOnRN } from 'react-native-worklets';

/** @internal */
export function updateSharedValWithTiming(
  sharedVal: SharedValue<number>,
  toVal: number,
  userConfig?: WithTimingConfig,
  cb?: () => void
) {
  sharedVal.set(
    withTiming(toVal, userConfig, () => {
      if (cb) {
        scheduleOnRN(cb);
      }
    })
  );
}

/** @internal */
export function updateSharedValWithSpring(
  sharedVal: SharedValue<number>,
  toVal: number,
  userConfig?: WithSpringConfig,
  cb?: () => void
) {
  sharedVal.set(
    withSpring(toVal, userConfig, () => {
      if (cb) {
        scheduleOnRN(cb);
      }
    })
  );
}
