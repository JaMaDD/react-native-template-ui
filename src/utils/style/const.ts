export enum Orientation {
  Portrait = 'portrait',
  Landscape = 'landscape',
}
