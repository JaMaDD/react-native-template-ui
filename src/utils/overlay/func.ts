import { use as reactUse } from 'react';
import { getWindowDimensionsHeight } from '../style/func';
import {
  ActionSheetContext,
  ActionSheetExpandableHeightRatio,
  actionSheetMaxHeightRatio,
  AlertContext,
} from './const';

export function getAlertContext() {
  return reactUse(AlertContext);
}

function getActionSheetExpandableHeightByRatio(
  ratio: ActionSheetExpandableHeightRatio
) {
  return getWindowDimensionsHeight() * ratio;
}

export function getActionSheetMaxHeight(expandable: boolean = false) {
  return expandable
    ? getActionSheetExpandableHeightByRatio(
        ActionSheetExpandableHeightRatio.Top
      )
    : getWindowDimensionsHeight() * actionSheetMaxHeightRatio;
}

export function getActionSheetHeight(height: number) {
  return Math.min(height, getActionSheetMaxHeight());
}

export function getActionSheetExpandableHeight() {
  return getActionSheetMaxHeight(true);
}

export function getActionSheetExpandableInitHeight() {
  return getActionSheetExpandableHeightByRatio(
    ActionSheetExpandableHeightRatio.Middle
  );
}

function getActionSheetDismiss(height: number, currentHeight: number) {
  return height / 2 < currentHeight;
}

function getActionSheetExpandableDismiss(currentHight: number) {
  return (
    currentHight <
    getActionSheetExpandableHeightByRatio(
      ActionSheetExpandableHeightRatio.Bottom
    ) /
      2
  );
}

export function getActionSheetSnapHeight(
  height: number,
  currentHeight: number,
  forceDismiss: boolean
) {
  return forceDismiss || getActionSheetDismiss(height, currentHeight)
    ? height
    : 0;
}

export function getActionSheetExpandableSnapHeight(
  currentHight: number,
  forceDismiss: boolean
) {
  if (forceDismiss || getActionSheetExpandableDismiss(currentHight)) {
    return 0;
  }

  const height = getWindowDimensionsHeight();
  const topHeight = getActionSheetExpandableHeightByRatio(
    ActionSheetExpandableHeightRatio.Top
  );
  const middleHeight = getActionSheetExpandableHeightByRatio(
    ActionSheetExpandableHeightRatio.Middle
  );
  const bottomHeight = getActionSheetExpandableHeightByRatio(
    ActionSheetExpandableHeightRatio.Bottom
  );
  if (currentHight > middleHeight) {
    return currentHight >
      (height *
        (ActionSheetExpandableHeightRatio.Top +
          ActionSheetExpandableHeightRatio.Middle)) /
        2
      ? topHeight
      : middleHeight;
  } else if (currentHight > bottomHeight) {
    return currentHight >
      (height *
        (ActionSheetExpandableHeightRatio.Middle +
          ActionSheetExpandableHeightRatio.Bottom)) /
        2
      ? middleHeight
      : bottomHeight;
  } else {
    return bottomHeight;
  }
}

export function getActionSheetContext() {
  return reactUse(ActionSheetContext);
}
