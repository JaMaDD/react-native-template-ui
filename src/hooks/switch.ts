import { useEffect } from 'react';
import {
  interpolate,
  interpolateColor,
  useAnimatedStyle,
  type SharedValue,
} from 'react-native-reanimated';
import type { PressableOnPress } from '../types/btn';
import type { AnimationValue } from '../types/reanimated';
import type { ThemedSwitchProps } from '../types/switch';
import { updateSharedValWithTiming } from '../utils/reanimated/func';
import { BorderSize, SwitchSize } from '../utils/theme/const';
import { useDelayedOnPress } from './btn';
import { useAnimationSharedVal } from './reanimated';
import { useThemeColors, useThemeSpacing } from './theme';

/** @internal */
export function useThemedSwitch(
  onPress: ThemedSwitchProps['onPress'],
  enabled: ThemedSwitchProps['enabled'] = false,
  size: ThemedSwitchProps['size'] = SwitchSize.M,
  onPressDelayConfig?: ThemedSwitchProps['onPressDelayConfig'],
  customEnableAnimation?: ThemedSwitchProps['customEnableAnimation'],
  customDisableAnimation?: ThemedSwitchProps['customDisableAnimation'],
  customColors?: ThemedSwitchProps['customColors']
) {
  const { xxxs } = useThemeSpacing();
  const { theme, background } = useThemeColors();
  const enabledSharedVal = useAnimationSharedVal();
  const pressableOnPress: PressableOnPress = () => {
    onPress(!!enabledSharedVal.get());
  };
  const delayedOnPress = useDelayedOnPress(
    pressableOnPress,
    onPressDelayConfig
  );
  const handleAnimation = (toVal: AnimationValue, cb?: () => void) => {
    if (customEnableAnimation && toVal) {
      customEnableAnimation(enabledSharedVal, toVal, cb);
    } else if (customDisableAnimation && !toVal) {
      customDisableAnimation(enabledSharedVal, toVal, cb);
    } else {
      updateSharedValWithTiming(
        enabledSharedVal as SharedValue<number>,
        toVal,
        { duration: 250 },
        cb
      );
    }
  };
  const toggleSwitch: PressableOnPress = (event) => {
    handleAnimation(enabledSharedVal.get() ? 0 : 1, () => {
      delayedOnPress(event);
    });
  };
  const trackAnimatedStyle = useAnimatedStyle(() => {
    const padding = xxxs;
    const borderWidth = BorderSize.S;
    const borderColor = theme;

    return {
      width: size * 2 + padding * 2 + borderWidth * 2,
      padding,
      borderWidth,
      borderColor: interpolateColor(
        enabledSharedVal.get(),
        [0, 1],
        [
          customColors?.border ?? borderColor,
          customColors?.borderEnabled ?? borderColor,
        ]
      ),
      backgroundColor: interpolateColor(
        enabledSharedVal.get(),
        [0, 1],
        [
          customColors?.background ?? background,
          customColors?.backgroundEnabled ?? borderColor,
        ]
      ),
    };
  }, [size, customColors, xxxs, theme, background, enabledSharedVal]);
  const thumbAnimatedStyle = useAnimatedStyle(
    () => ({
      backgroundColor: interpolateColor(
        enabledSharedVal.get(),
        [0, 1],
        [customColors?.thumb ?? theme, customColors?.thumbEnabled ?? background]
      ),
      transform: [
        { translateX: interpolate(enabledSharedVal.get(), [0, 1], [0, size]) },
      ],
    }),
    [size, customColors, theme, background, enabledSharedVal]
  );
  useEffect(() => {
    handleAnimation(enabled ? 1 : 0);
  }, [enabled]);

  const switchProps = {
    toggleSwitch,
    trackAnimatedStyle,
    thumbAnimatedStyle,
  };

  return switchProps;
}
