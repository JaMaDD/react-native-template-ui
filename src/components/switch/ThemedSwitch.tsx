import type { FC } from 'react';
import { useThemedSwitch } from '../../hooks/switch';
import type { ThemedSwitchProps } from '../../types/switch';
import { composeStyles, flattenStyle } from '../../utils/style/func';
import { SwitchSize } from '../../utils/theme/const';
import AnimatedThemedPressable from '../btn/AnimatedThemedPressable';
import AnimatedThemedView from '../view/AnimatedThemedView';

const ThemedSwitch: FC<ThemedSwitchProps> = ({
  onPress,
  enabled,
  size = SwitchSize.M,
  onPressDelayConfig,
  customEnableAnimation,
  customDisableAnimation,
  customColors,
  thumbProps,
  style,
  ...props
}) => {
  const { toggleSwitch, trackAnimatedStyle, thumbAnimatedStyle } =
    useThemedSwitch(
      onPress,
      enabled,
      size,
      onPressDelayConfig,
      customEnableAnimation,
      customDisableAnimation,
      customColors
    );

  const switchStyle = flattenStyle(composeStyles(style, trackAnimatedStyle));

  return (
    <AnimatedThemedPressable
      onPress={toggleSwitch}
      style={switchStyle}
      role={'switch'}
      {...props}
    >
      <AnimatedThemedView
        width={size}
        aspectRatio={1}
        style={thumbAnimatedStyle}
        {...thumbProps}
      />
    </AnimatedThemedPressable>
  );
};

export default ThemedSwitch;
