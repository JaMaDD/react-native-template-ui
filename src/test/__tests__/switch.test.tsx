import type { FC } from 'react';
import ThemedSwitch from '../../components/switch/ThemedSwitch';
import type { ThemedSwitchProps } from '../../types/switch';
import { switchAnimationDuration } from '../../utils/switch/const';
import { getThemeColors, hexToRgb } from '../utils/func';
import { renderAsync, screen, userEvent } from '../utils/testingLib';

const Switch: FC<Omit<ThemedSwitchProps, 'onPress'>> = ({
  thumbProps,
  ...props
}) => (
  <ThemedSwitch
    onPress={() => {}}
    thumbProps={{ ...thumbProps, testID: 'switchThumb' }}
    {...props}
  />
);

function getSwitch() {
  return screen.getByRole('switch');
}

function getSwitchThumb() {
  return screen.getByTestId('switchThumb');
}

describe('ThemedSwitch', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  test('Render & Snapshot', async () => {
    const { toJSON } = await renderAsync(<Switch />);
    expect(getSwitch()).toBeOnTheScreen();
    expect(toJSON()).toMatchSnapshot();
  });

  test('Press with default styles', async () => {
    const { theme, background } = await getThemeColors();
    await renderAsync(<Switch />);
    const switchComponent = getSwitch();
    const switchThumb = getSwitchThumb();
    expect(switchComponent).toHaveAnimatedStyle({
      backgroundColor: hexToRgb(background),
    });
    expect(switchThumb).toHaveAnimatedStyle({
      backgroundColor: hexToRgb(theme),
    });
    const user = userEvent.setup();
    await user.press(switchComponent);
    jest.advanceTimersByTime(switchAnimationDuration + 50);
    expect(switchComponent).toHaveAnimatedStyle({
      backgroundColor: hexToRgb(theme),
    });
    expect(switchThumb).toHaveAnimatedStyle({
      backgroundColor: hexToRgb(background),
    });
  });

  test('Press with custom styles', async () => {
    const { theme, background } = await getThemeColors();
    await renderAsync(
      <Switch
        customColors={{
          background: 'theme',
          backgroundEnabled: 'background',
          border: 'background',
          borderEnabled: 'background',
          thumb: 'background',
          thumbEnabled: 'theme',
        }}
      />
    );
    const switchComponent = getSwitch();
    const switchThumb = getSwitchThumb();
    expect(switchComponent).toHaveAnimatedStyle({
      borderColor: hexToRgb(background),
      backgroundColor: hexToRgb(theme),
    });
    expect(switchThumb).toHaveAnimatedStyle({
      backgroundColor: hexToRgb(background),
    });
    const user = userEvent.setup();
    await user.press(switchComponent);
    jest.advanceTimersByTime(switchAnimationDuration + 50);
    expect(switchComponent).toHaveAnimatedStyle({
      borderColor: hexToRgb(background),
      backgroundColor: hexToRgb(background),
    });
    expect(switchThumb).toHaveAnimatedStyle({
      backgroundColor: hexToRgb(theme),
    });
  });

  test('Change Enabled prop', async () => {});

  test('Press with delay config', async () => {});
});
