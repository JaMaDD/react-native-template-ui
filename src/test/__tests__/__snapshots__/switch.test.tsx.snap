// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemedSwitch Render & Snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    jestAnimatedProps={
      {
        "value": {},
      }
    }
    jestAnimatedStyle={
      {
        "value": {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderColor": "rgba(124, 207, 107, 1)",
          "borderWidth": 1,
          "padding": 2,
          "width": 50,
        },
      }
    }
    jestInlineStyle={
      [
        {},
      ]
    }
    nativeID="0"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="switch"
    style={
      [
        {},
        {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderColor": "rgba(124, 207, 107, 1)",
          "borderWidth": 1,
          "padding": 2,
          "width": 50,
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      jestAnimatedProps={
        {
          "value": {},
        }
      }
      jestAnimatedStyle={
        {
          "value": {
            "backgroundColor": "rgba(124, 207, 107, 1)",
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      jestInlineStyle={{}}
      nativeID="1"
      role="group"
      style={
        [
          {
            "aspectRatio": 1,
            "backgroundColor": "#ffffff",
            "width": 22,
          },
          [
            {
              "backgroundColor": "rgba(124, 207, 107, 1)",
              "transform": [
                {
                  "translateX": 0,
                },
              ],
            },
          ],
        ]
      }
      testID="switchThumb"
    />
  </View>
</View>
`;
